src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+
+  if (!get_upgrade_info_from_branch_file (newest,
+                                          &refspec,
+                                          &original_refspec,
+                                          &metrics,
+                                          error))
+    return FALSE;
+
+  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
+    return FALSE;
+
+  swbfs_with_latest_commit = object_array_new ();
+  urls = g_ptr_array_new_with_free_func (g_free);
+  for (idx = 0; idx < swbfs->len; ++idx)
+    {
+      gpointer swbf_ptr = g_ptr_array_index (swbfs, idx);
+      EosServiceWithBranchFile *swbf = EOS_SERVICE_WITH_BRANCH_FILE (swbf_ptr);
+      EosAvahiService *service = swbf->service;
+      g_autofree gchar *url_override = NULL;
+      g_autoptr(GError) local_error = NULL;
+      g_autofree gchar *checksum = NULL;
+      g_autoptr(GVariant) commit = NULL;
+      guint64 timestamp;
+      g_autoptr(EosExtensions) extensions = NULL;
+      GCancellable *cancellable;
+
+      cancellable = g_task_get_cancellable (lan_data->fetch_data->task);
+      url_override = g_strdup_printf ("http://%s:%d",
+                                      service->address,
+                                      service->port);
+      if (!fetch_latest_commit (repo,
+                                cancellable,
+                                remote,
+                                ref,
+                                url_override,
+                                &checksum,
+                                &extensions,
+                                &local_error))
+        {
+          message ("Failed to fetch latest commit from %s: %s",
+                   url_override,
+                   local_error->message);
+          continue;
+        }
+
+      if (!is_checksum_an_update (repo, checksum, &commit, &local_error))
+        {
+          message ("Commit %s from %s is not an update",
+                   checksum,
+                   url_override);
+          continue;
+        }
+
+      timestamp = ostree_commit_get_timestamp (commit);
+      if (latest_checksum != NULL)
+        {
+          if (timestamp < latest_timestamp)
+            continue;
+
+          if (timestamp == latest_timestamp && g_strcmp0(checksum, latest_checksum) == 0)
+            {
+              g_ptr_array_add (swbfs_with_latest_commit, g_object_ref (swbf));
+              g_ptr_array_add (urls, g_steal_pointer (&url_override));
+            }
+          else if (timestamp > latest_timestamp && g_strcmp0(checksum, latest_checksum) != 0)
+            {
+              g_clear_pointer (&latest_checksum, g_free);
+              g_clear_pointer (&latest_commit, g_variant_unref);
+              latest_timestamp = 0;
+              g_ptr_array_set_size (swbfs_with_latest_commit, 0);
+              g_ptr_array_set_size (urls, 0);
+              g_clear_object (&latest_extensions);
+            }
+        }
+      if (latest_checksum == NULL)
```

Comment:
this might add swbfs and url_override to swbfs_with_latest_commit again and urls if both latest_checksum and checksum are NULL, since g_strcmp0 considers two NULLs to be equal.